### ISSUE TITLE
Remove labels with insufficient data during pre-processing

### ISSUE THREAD

**Issue by paul-tqh-nguyen** [paul-tqh-nguyen/reuters_topic_labelling#18]
Useful Code Snippet:
```
from main import *
from  models import *

classifier = EEAPClassifier(NUMBER_OF_EPOCHS, BATCH_SIZE, TRAIN_PORTION, VALIDATION_PORTION, TESTING_PORTION, MAX_VOCAB_SIZE, PRE_TRAINED_EMBEDDING_SPECIFICATION, ENCODING_HIDDEN_SIZE, NUMBER_OF_ENCODING_LAYERS, ATTENTION_INTERMEDIATE_SIZE, NUMBER_OF_ATTENTION_HEADS, DROPOUT_PROBABILITY, OUTPUT_DIR)

all_examples = [e for e in classifier.all_data]
all_data_label_histogram = torch.tensor([[bool(getattr(example, topic)) for topic in classifier.topics] for example in all_examples], dtype=int).sum(dim=0)

training_examples = [e for e in classifier.training_data]
training_data_label_histogram = torch.tensor([[bool(getattr(example, topic)) for topic in classifier.topics] for example in training_examples], dtype=int).sum(dim=0)

validation_examples = [e for e in classifier.validation_data]
validation_data_label_histogram = torch.tensor([[bool(getattr(example, topic)) for topic in classifier.topics] for example in validation_examples], dtype=int).sum(dim=0)

testing_examples = [e for e in classifier.testing_data]
testing_data_label_histogram = torch.tensor([[bool(getattr(example, topic)) for topic in classifier.topics] for example in testing_examples], dtype=int).sum(dim=0)

label_counts_too_small_to_reach_training_data = sorted([e.item() for e in all_data_label_histogram*((training_data_label_histogram == 0).int()) if e.item() != 0])
label_counts_too_small_to_reach_validation_data = sorted([e.item() for e in all_data_label_histogram*((validation_data_label_histogram == 0).int()) if e.item() != 0])
label_counts_too_small_to_reach_testing_data = sorted([e.item() for e in all_data_label_histogram*((testing_data_label_histogram == 0).int()) if e.item() != 0])
```
Exploring the data has revealed that there are some labels that have too few samples to even have one sample be given to the training, validation, and testing set. 

Thus, there are some labels that our model CANNOT EVER learn. 

We should get rid of these from our dataset during pre-processing. 

It should be easy to do once our data is in a `pandas.DataFrame`.

Let's leave the `all_extracted_data.csv` file alone since it is intended to contain everything and just modify how we generate `topics_data.csv`.

Here's some important data:
```
>>> label_counts_too_small_to_reach_training_data
[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2]
>>> label_counts_too_small_to_reach_validation_data
[1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 5]
>>> label_counts_too_small_to_reach_testing_data
[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 4, 4, 6]
>>> 
```
Let's set the threshold of minimum number of required datapoints to be 10 and see how well our model performs.

**Comment by paul-tqh-nguyen** [comment]
https://github.com/paul-tqh-nguyen/reuters_topic_labelling/commit/005fc70afe2269e02415871b753c55ac7aea02d0 removes topics with insufficiently many samples (must have at least 10. 

Next and final step is to filter out rows who have no topics.

**Comment by paul-tqh-nguyen** [comment]
https://github.com/paul-tqh-nguyen/reuters_topic_labelling/commit/c6048c85662214589ac30defe4e47a9fc387ccb0 This patch completes all the tasks for this ticket. 

We have the benefit that we now converge to an F1 score of 0.8 in ~3 epochs. That's a huge improvement.
